### PR TITLE
bgfx crt-geom and crt-geom-deluxe enhancements

### PR DESCRIPTION
**Pull request by cgwg**
Contains three changes to these shaders:
1. A "brightness boost" feature for the shadow mask that works by making the brightness ratio between bright and dark mask pixels closer to 1 for the brighter parts of the image. I've been careful to ensure that the average of bright and dark mask pixels still varies linearly, so that the tone curve is unaffected as long as the gamma of the output display is set correctly.
2. Clamping to zero so that underscanning produces a black border.
3. In crt-geom-deluxe, a "raster bloom" effect that makes the image grow slightly when the average brightness of the screen is high. This mimics a common defect in CRTs and has been previously implemented in the shader called crt-guest-dr-venom. I find the effect kind of nice when the image fades in from black.

One detail I am uncertain about comes from the brightness boost feature: this required encoding some extra information in the alpha channel of the shadow mask textures. I wrote a very simple Python script to do this and have included this script, but perhaps this is inappropriate or it should be put in a different directory.

Finally: the shaders have only been compiled for glsl; someone else will have to generate the full set of shader targets.

### CODE REVIEW

**Comment by john-iv** [comment]
This sounds exciting, remedying one of the issues w/ the crt-geom chains being too dark.  I've been enjoying your LCD-Grid effect as my new default for playing; do these changes positively impact lcd-grid too or are they diminishing returns?

**Comment by cgwg** [comment]
There's nothing here for lcd-grid. The general tradeoff is that any pattern superimposed on the image will make it darker, and tricks to counteract this will weaken the pattern. An additional difficulty for such a trick is to avoid unintentionally altering the tone curve. I'm not currently planning any such changes to lcd-grid.

File: bgfx/effects/crt-geom/mipmap8.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    	"alpha": true
+    },
+    "vertex": "chains/crt-geom/vs_mipmap8",
+    "fragment": "chains/crt-geom/fs_mipmap8",
+    "uniforms": [
+	{ "name": "u_tex_size0",     "type": "vec4", "values": [ 1.0, 1.0, 0.0, 0.0 ] },
+	{ "name": "u_tex_size1",     "type": "vec4", "values": [ 1.0, 1.0, 0.0, 0.0 ] },
+	{ "name": "u_smooth",        "type": "vec4", "values": [ 0.5, 0.0, 0.0, 0.0 ] },
+	{ "name": "s_screen",     "type": "int",  "values": [ 0 ] },
+	{ "name": "s_mipmap",     "type": "int",  "values": [ 1 ] }
```
Comment (cuavas) [inline bgfx/effects/crt-geom/mipmap8.json:27]: Something’s gone wrong with indentation here.

**Comment by cgwg** [comment]
How embarrassing. I've now tried to clean up all of the whitespace and indentation to make it consistent.

**Review by cuavas — COMMENTED** [review]
Why is github showing 0-byte PNG files added?  This seems a little suspect.

**Comment by u-man74** [comment]
I dont know why github says that they are "empty files", but if you say "view file" under the "..." menue in github, they will show up properly. Maybe this is because the files are really tiny in size.

**Comment by cuavas** [comment]
Weird, another GitHub UI bug.

**Comment by cuavas** [comment]
Shaders compiled for the other back-ends in 574b21ffb971ce96f6390b0567ff5a1f2ea5a992.  Works on Windows with most back-ends, but `d3d9` seems to be broken at the moment (not specific to this, all the screen chains I tested gave a black screen with `-bgfx_backend d3d9`).

**Comment by ICEknigh7** [comment]
> 3. In crt-geom-deluxe, a "raster bloom" effect that makes the image grow slightly when the average brightness of the screen is high. This mimics a common defect in CRTs and has been previously implemented in the shader called crt-guest-dr-venom. I find the effect kind of nice when the image fades in from black.
Was this implemented in the end? The intended effect should be noticeable with the glowing title logo and level titles in genesis[oozeju], but the screen size doesn't seem to change even with the settings at their maximum values.

**Comment by cgwg** [comment]
@ICEknigh7 Yes, it was added and it works for me using mame 0.230 on Fedora with Intel graphics. The one issue that I could imagine causing problems is that the mipmap shader reads from and writes to the same texture. Can you try running this test shader chain (rename it to test.json and put it in bgfx/chains)? You should see two or three scaled-down versions of the screen, and the right edge should show the average colour.

[test.txt](https://github.com/mamedev/mame/files/6292802/test.txt)

**Comment by ICEknigh7** [comment]
@cgwg I'm just getting a tiny screen on the top left corner: https://i.imgur.com/bNJ5QPP.png

**Comment by cgwg** [comment]
@ICEknigh7 OK, that's not working. Try this alternative test and alternative crt-geom-deluxe.
[crt-geom-deluxe.txt](https://github.com/mamedev/mame/files/6305134/crt-geom-deluxe.txt)
[test2.txt](https://github.com/mamedev/mame/files/6305135/test2.txt)

**Comment by ICEknigh7** [comment]
@cgwg Nothing seems to change. 😕

https://i.imgur.com/zBd3gQL.png

**Comment by ICEknigh7** [comment]
Something else that I've noticed is that the "halation" effect makes crt-geom-deluxe look more like a picture taken off a CRT with an out of focus lens, rather than how it looks in person.

Here's a real-life example with high black/white contrasts (although the pic still came out just a bit brighter than it should):
- Current crt-geom-deluxe with default values: https://i.imgur.com/a5lPVO9.png
- Real CRT: https://i.imgur.com/ihgeQlP.jpg

Maybe something's wrong with the calculations, I'm not sure that it should be left on by default just yet.

**Comment by cgwg** [comment]
@ICEknigh7 Unfortunately it's difficult to debug the mipmap shader when everything works on my setup. The other feature that requires a texture to be preserved between frames is the phosphor persistence. Does that work for you? The strongest (very obvious) effect can be produced by setting "Phosphor persistence amplitude" to its maximum and "Phosphor decay power" to its minimum.

In your photo the halation is longer distanced and has a smaller amplitude than the shader default. I agree that the default effect is near the upper end of what is reasonable. Making it much longer ranged would require adjusting the shader (adding more taps to the filter).

**Comment by ICEknigh7** [comment]
@cgwg 
> The other feature that requires a texture to be preserved between frames is the phosphor persistence. Does that work for you? The strongest (very obvious) effect can be produced by setting "Phosphor persistence amplitude" to its maximum and "Phosphor decay power" to its minimum.

Yeah, phosphor persistance works fine on my end.

**Comment by cgwg** [comment]
@ICEknigh7 Sorry for the slow reply; I haven't forgotten about this.
Unfortunately it's not easy to debug shaders, and doing it remotely is even worse. To start: maybe it would be useful to know what video hardware you have and which bgfx backend you are using. (Have you tried different backends?)

**Comment by ICEknigh7** [comment]
I have an Nvidia GeForce 630N.

The raster bloom effect works on:
-opengl

Doesn't work on:
-d3d11
-d3d9

**Comment by ICEknigh7** [comment]
@cgwg Regarding how the effect mimics that of a real CRT, I've noticed that there might be something off with the calculations, since the Title Screen of Sonic 3D Blast shouldn't feature noticeable bloom for some reason:
https://youtu.be/liKTg16kKgg

(The text on the intro does show the effect normally, on the other hand)

**Comment by cgwg** [comment]
@ICEknigh7 Thanks for the information. Unfortunately I don't have access to a Windows setup, so I can't debug d3d9 or d3d11.
The video examples are very nice. My guess about the Sonic 3D Blast example is that this effect is acting very quickly, i.e. not lasting much beyond one frame. My simulation expects things to change a bit slower, and there's even a minimum lag of a few frames. It would fairly simple to reduce the lag by adding more passes.
In general I expect it to vary between CRTs, but it would be interesting to see how quickly this effect can act: can it be seen within a few scanlines? If I had the time and resources, it would be a fun project to try to really model the behaviour of a particular display.

**Comment by ICEknigh7** [comment]
Right now (from my limited knowledge), I think the only feature truly missing from this "raster bloom" effect might be the "bounce" which can also be seen in this monitor as well:

kunio dodge ball neo geo

This is indeed a really interesting and great-looking effect, which I personally think should be on by default when polished enough (that's how CRTs look, after all). Perhaps it would be a good topic to bring up in MAME's forums (or Discord, even), where more examples of real hardware can be posted to help the research.

**Comment by ICEknigh7** [comment]
@cgwg In the meantime, do you think it would be possible to expand and adjust the "halation" effect to something more similar to what's seen on the real-hardware Flicky example? The current default values for it can be a bit eye-straining.

**Comment by ICEknigh7** [comment]
@cgwg  Do you have a place for posting bug reports and real hardware reference?

I opened a bug report at MameTesters but got closed due to crt-geom-deluxe being third party: [https://mametesters.org/view.php?id=8089](url)

**Pull request merged**